docs: Overhaul Explore Profiles documentation (#63)
grafana/explore-profiles: docs/sources/concepts/index.md

---
description: Learn about concepts basic to profiling.
keywords:
  - Explore Profiles
  - Concepts
title: Concepts
menuTitle: Concepts
weight: 200
---

# Concepts

To use profiling, you should understand these concepts:

* Profile data hierarchy
* Profile types
* Labels
* Flame graphs

## Profile data hierarchy

Each profile has data for services. A service can be one of the processes running on your system or application.

> Each service has multiple profile types, such as CPU, memory, exceptions, and locks. These profiles provide details about processes within a service.

Reviewer: ```suggestion
Each service has multiple profile types, such as CPU, memory, exceptions, and locks. These profiles provide details about the processes in a service.
```

> * Service:  This is the main application or system you are monitoring. For example, it could be an online shopping platform, a social media application, or a cloud storage service

Reviewer: ```suggestion
* Service:  This is the main application or system you are monitoring. For example, it could be an online shopping platform, a social media application, or a cloud storage service.
```

>   * Profile type: Each service can be monitored through different profile types, each focusing on a specific performance aspect. Example profile types:

Reviewer: ```suggestion
  * Profile type: You can monitor each service through different profile types, each focusing on a specific performance aspect. Example profile types:
```

>     * Cpu

Reviewer: ```suggestion
    * CPU
```

For more information, refer to [Understand profiling types](https://grafana.com/docs/grafana-cloud/monitor-applications/profiles/introduction/profiling-types/).

<!-- vale Grafana.Spelling = NO -->
| Profile type | What it shows | When to use |
|---|---|---|
| CPU profiling | Measures the amount of CPU time consumed by different parts of your application code | To identify and optimize CPU-intensive functions |
| Memory allocation  | Tracks the amount and frequency of memory allocations by the application | For identifying and optimizing memory usage patterns |
| Goroutine | Measures the usage and performance of lightweight routines in Go | Especially useful in Go applications for concurrency management |
| Contentions (Mutex) | Mutex profiling involves analyzing mutex (mutual exclusion) locks, used to prevent simultaneous access to shared resources. | To optimize thread synchronization and reduce lock contention |
| Block | Measures the frequency and duration of blocking operations, where a thread is paused or delayed | To identify and reduce blocking delays |
<!-- vale Grafana.Spelling = NO -->

## Labels

> Labels help identify parts of profiling data. For example, labels identify Kubernetes Pod names, namespaces, repositories, references, and more.

Reviewer: ```suggestion
Labels help identify parts of profiling data. For example, labels identify Kubernetes Pod names, namespaces, repositories, and references.
```

Reviewer: I feel like if there is more, we should specify those things here.

Author: There are a LOT more. You can create a label for almost anything using custom labels.

This feature is crucial for identifying performance anomalies and understanding the behavior of different application segments under various conditions.

![Labels view in Explore Profiles](../images/explore-profiles-labels.png)

## Flame graphs

> A fundamental aspect of continuous profiling is the flame graph, a convenient way to visualize performance data. These graphs provide a clear, intuitive understanding of resource allocation and bottlenecks within the application.

Reviewer: ```suggestion
A fundamental aspect of continuous profiling is the flame graph, a convenient way to visualize performance data. These graphs provide a clear, intuitive understanding of resource allocation and bottlenecks in the application.
```

Reviewer: A manager I once had told me "within --> in" and it stuck with me 😄
